From: SeongJae Park <sj@kernel.org>
Date: Fri, 13 Dec 2024 11:40:49 -0800
Subject: [PATCH] mm/damon/sysfs: do scheme region directories removal on sysfs
 context

The work has no real reason to be done on kdamond context.  Do it on
sysfs context.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/sysfs.c | 21 ++++++++-------------
 1 file changed, 8 insertions(+), 13 deletions(-)

diff --git a/mm/damon/sysfs.c b/mm/damon/sysfs.c
index 8e8957cbf52f..fa494c48c6f6 100644
--- a/mm/damon/sysfs.c
+++ b/mm/damon/sysfs.c
@@ -1468,16 +1468,6 @@ struct damon_sysfs_schemes_walk_data {
 	bool total_bytes_only;
 };
 
-/* cleanup regions directories */
-static void damon_sysfs_schemes_tried_regions_upd_prep(void *data, struct damon_ctx *ctx)
-{
-	struct damon_sysfs_schemes_walk_data *walk_data = data;
-	struct damon_sysfs_kdamond *sysfs_kdamond = walk_data->sysfs_kdamond;
-
-	damon_sysfs_schemes_clear_regions(
-			sysfs_kdamond->contexts->contexts_arr[0]->schemes, ctx);
-}
-
 /* poppulate the region directory */
 static void damon_sysfs_schemes_tried_regions_upd_one(void *data, struct damon_ctx *ctx,
 		struct damon_target *t, struct damon_region *r,
@@ -1499,14 +1489,19 @@ static int damon_sysfs_update_schemes_tried_regions(
 		.total_bytes_only = total_bytes_only,
 	};
 	struct damos_walk_control control = {
-		.prep_fn = damon_sysfs_schemes_tried_regions_upd_prep,
+		.prep_fn = NULL,
 		.walk_fn = damon_sysfs_schemes_tried_regions_upd_one,
 		.data = &walk_data,
 	};
+	struct damon_ctx *ctx = sysfs_kdamond->damon_ctx;
 
-	if (!sysfs_kdamond->damon_ctx)
+	if (!ctx)
 		return -EINVAL;
-	return damos_walk(sysfs_kdamond->damon_ctx, &control);
+
+	damon_sysfs_schemes_clear_regions(
+			sysfs_kdamond->contexts->contexts_arr[0]->schemes,
+			ctx);
+	return damos_walk(ctx, &control);
 }
 
 /*
-- 
2.39.5

